### PR TITLE
Put cockpit-ws in a container

### PR DESCRIPTION
**Pull request by stefwalter**
For Fedora Atomic, we want cockpit-ws to be able to run in a container, with the cockpit-bridge and friends running on the host.

We add code to cockpit-session to do the 'nsenter' style namespace switch for us at the appropriate place. This means that cockpit-session doesn't need to be in the host, but can still run the host PAM stack and so on.

We change around our Dockerfile to reflect this new approach.

### CODE REVIEW

**Comment by stefwalter** [comment]
@cgwalters This is relevant to your interests :)

@rhatdan The PAM stack currently fails on SELinux after we make the nsenter switch and drop into the host namespaces. We get the following message from pam_selinux.so in the host.

```
pam: Unable to get valid context for stef
```

And our login fails due to SELinux failures.

**Comment by stefwalter** [comment]
Should we use readlink() and make ```/container/target-namespace``` be a symlink?

File: tools/Dockerfile [hunk 1]
```diff
@@ -3,7 +3,7 @@ MAINTAINER "Stef Walter" <stefw@redhat.com>
```
Comment (dperpeet) [inline tools/Dockerfile]: fedora:21 explicitly or just fedora:latest?

File: src/ws/session.c [hunk 2]
```diff
@@ -78,17 +82,19 @@ read_auth_until_eof (size_t *out_len)
       if (alloc <= len)
         {
           alloc += 1024;
+          if (alloc > 64 * 1024)
```
Comment (dperpeet) [inline src/ws/session.c]: should we have a define at the top of the file / somewhere central for limits like these?
at least document the hard limit for the function

File: src/ws/session.c [hunk 3]
```diff
@@ -78,17 +82,19 @@ read_auth_until_eof (size_t *out_len)
...
+            errx (EX, "input data is too long %s", what);
```
Comment (dperpeet) [inline src/ws/session.c]: style of message inconsistent with allocate error: ... `too long for %s` would be consistent

File: tools/Dockerfile (hunk 1 above)
Comment (stefwalter) [inline tools/Dockerfile]: fedora:22 doesn't exist yet ... and given the wild state of Fedora, God knows what would get pushed in there ... lets stick with what we know ... and only update when after smoke testing a new base image.

File: tools/Dockerfile [hunk 4]
```diff
@@ -3,21 +3,19 @@ MAINTAINER "Stef Walter" <stefw@redhat.com>
...
 
 # If there are rpm files in the current directory we'll install those,
 # otherwise use cockpit-preview repo. The Dockerfile is a hack around
 # Dockerfile lack of support for branches
-ADD *.rpm Dockerfile /tmp/
+ADD cockpit-ws-*.rpm Dockerfile /tmp/
 
 # Again see above ... we do our branching in shell script
-RUN cd /tmp && ( ls *.rpm > /dev/null 2> /dev/null && yum -y install cockpit*.rpm || yum -y --enablerepo=cockpit-preview install cockpit ) && yum clean all && rm -f /tmp/cockpit*.rpm
+RUN cd /tmp && ( ls *.rpm > /dev/null 2> /dev/null && yum -y install *.rpm || yum -y --enablerepo=cockpit-preview install cockpit-ws ) && yum clean all && rm -f /tmp/*.rpm
```
Comment (dperpeet) [inline tools/Dockerfile]: we should split this into multiple lines
https://docs.docker.com/articles/dockerfile_best-practices/

**Comment by stefwalter** [comment]
Tested again. Works well enough to merge (see above caveat). Added suggested changes from @dperpeet to Dockefile and are now using ```readlink()```

**Comment by dperpeet** [comment]
looks good

**Comment by rhatdan** [comment]
Stef what pam stack are you using?  Do we actually want to use a user context for this?  The problem is that we don;t have a transition defined between the context running the login/su and a user domain.

**Pull request closed without merging**